float32 distance
float32 height
float32 angle